### PR TITLE
Feat/v1.3: Refactor for efficiency and speed

### PR DESCRIPTION
**Pull request by parambharat**
This PR includes the following changes to improve the efficiency and speed of wandbot

- Replaces llama-index with LECL and makes LLM calls parallel
- Improves parsing of documents in the ingestion
- Replaces FAISS with ChromaDB
- Improves the query enhancer to be a single LLM call
- Splits the RAG pipeline into query enhancement, retrieval, and response synthesis modules
- Adds parent document retrieval in the retrieval module
- Adds sub-query answering in the response synthesis
- Splits the API into routers for retrieval, database and chat - @ Breaking change for API definitions

TODO: Need to change the eval modules to use LECL from llama-index  - @ayulockin

### CODE REVIEW

File: src/wandbot/ingestion/config.py [hunk 1]
```diff
@@ -224,9 +241,8 @@ def _set_cache_paths(cls, values: "DataStoreConfig") -> "DataStoreConfig":
 
 class VectorStoreConfig(BaseSettings):
     name: str = "vectorstore"
-    embedding_dim: int = 1536
+    embeddings_model: str = "text-embedding-3-small"
```
Comment (morganmcg1) [inline src/wandbot/ingestion/config.py:244]: lets make this the large model? unless there is a dramatic increase in latency? but an extra few seconds is fine if it means a higher chance of better retrieval

File: src/wandbot/ingestion/prepare_data.py [hunk 2]
```diff
@@ -778,7 +762,7 @@ def extract_tags(source_url: str, report_content: str) -> List[str]:
         tags = list(set([tag.title() for tag in tags]))
 
         if ("wandb.log" or "wandb.init") in report_content:
-            tags.append("contains-wandb-code")
+            tags.append("Contains Wandb Code")
```
Comment (morganmcg1) [inline src/wandbot/ingestion/prepare_data.py:765]: can we keep these tags as lowercased, hyphenated? it'll help with passing tags via endpoints i think

File: src/wandbot/ingestion/prepare_data.py [hunk 3]
```diff
@@ -790,19 +774,16 @@ def lazy_load(self) -> Iterator[Document]:
...
                 metadata={
                     "source": parsed_row["source"],
-                    "language": lang_detect.detect_language(
-                        parsed_row["content"]
-                    ),
+                    "source_type": self.config.source_type,
                     "file_type": ".md",
                     "description": parsed_row["description"],
-                    "tags": ["fc-report"]
+                    "tags": ["Fully Connected", "Report"]
```
Comment (morganmcg1) [inline src/wandbot/ingestion/prepare_data.py:786]: same lowercased, hyphenated comment as above. also lets refer to 'reports' as articles from now

File: src/wandbot/ingestion/prepare_data.py [hunk 4]
```diff
@@ -103,12 +102,12 @@ def _get_local_paths(self):
             )
 
         local_paths = []
-        file_patterns = (
-            [self.config.data_source.file_pattern]
-            if isinstance(self.config.data_source.file_pattern, str)
-            else self.config.data_source.file_pattern
-        )
-        for file_pattern in file_patterns:
+        # file_patterns = (
```
Comment (morganmcg1) [inline src/wandbot/ingestion/prepare_data.py:105]: keeping this code or?

File: src/wandbot/ingestion/prepare_data.py [hunk 5]
```diff
@@ -838,47 +846,26 @@ def load(
...
+        SDKCodeStoreConfig(),
+        SDKTestsStoreConfig(),
+        WeaveCodeStoreConfig(),
+        WeaveExamplesStoreConfig(),
+        WandbEduCodeStoreConfig(),
+        FCReportsStoreConfig(),
+    ]
+
+    pool = Pool(cpu_count() - 1)
+    results = pool.imap_unordered(load_from_config, configs)
```
Comment (morganmcg1) [inline src/wandbot/ingestion/prepare_data.py:863]: nice

File: src/wandbot/ingestion/preprocess_data.py [hunk 6]
```diff
@@ -19,146 +19,210 @@
...
-    special_tokens_set = encoding.special_tokens_set
+    def decode(self, tokens):
+        return self.tokenizer.decode(tokens)
 
-    def remove_special_tokens(text: str) -> str:
-        """Removes special tokens from the given text.
 
-        Args:
-            text: A string representing the text.
+tokenizer = Tokenizer("gpt2")
```
Comment (morganmcg1) [inline src/wandbot/ingestion/preprocess_data.py:53]: shouldn't we use the 3.5/4 tokenizer?

File: src/wandbot/ingestion/preprocess_data.py [hunk 7]
```diff
@@ -19,146 +19,210 @@
...
-                parser = code_parser
-            else:
-                parser = md_parser
-            nodes.extend(parser.get_nodes_from_documents([doc]))
-        except Exception as e:
-            logger.error(f"Error parsing document: {e}")
-            logger.warning(
-                f"Unable to parse document: {doc.metadata['source']} with custom parser, using default "
-                f"parser instead."
+    lang_detect = FastTextLangDetect()
```
Comment (morganmcg1) [inline src/wandbot/ingestion/preprocess_data.py:179]: i thought you mentioned we were removing langdetect? or is that the next update?

File: src/wandbot/ingestion/vectorstores.py [hunk 8]
```diff
@@ -55,45 +51,47 @@ def load(
...
     )
     artifact_dir: str = artifact.download()
-    storage_context = load_storage_context(config.embedding_dim)
-    service_context = load_service_context(
-        embeddings_cache=str(config.embeddings_cache),
-        llm="gpt-3.5-turbo-16k-0613",
-        temperature=config.temperature,
-        max_retries=config.max_retries,
+
+    # Todo: Change to LiteLLM Embeddings
```
Comment (morganmcg1) [inline src/wandbot/ingestion/vectorstores.py:62]: is this TODO for the next PR?

File: src/wandbot/rag/query_handler.py [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+    "You are given a conversation and a follow-up query. "
+    "You goal to enhance the user query and render it using the tool provided."
+    "\n\nChat History: \n\n"
+    "{chat_history}"
+)
+
+ENHANCER_PROMPT_MESSAGES = [
+    ("system", ENHANCER_SYSTEM_PROMPT),
+    ("human", "Question: {query}"),
+    ("human", "!!! Tip: Make sure to answer in the correct format"),
```
Comment (morganmcg1) [inline src/wandbot/rag/query_handler.py:254]: lol

File: src/wandbot/rag/response_synthesis.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
...an use the `wandb.Audio()` to create audio instances and log them to W&B using `wandb.log()`.\n\nLog an audio array or file\nwandb.log({{"my whale song": wandb.Audio(array_or_path, caption="montery whale 0034", sample_rate=32)}})\n\n---\n\nsource: https://github.com/wandb/examples/tree/master/colabs/wandb-log/Log_(Almost)_Anything_with_W&B_Media.ipynb\n\nLog multiple audio files\nLog audio within a W&B Table\n\nmy_table = wandb.Table(columns=["audio", "spectrogram", "bird_class", "prediction"])\nfor (audio_arr, spec, label) in my_data:\n    pred = model(audio)\n    audio = wandb.Audio(audio_arr, sample_rate=32)\n    img = wandb.Image(spec)\n    my_table.add_data(audio, img, label, pred)\n\nLog the Table to wandb\nwandb.log({{"validation_samples" : my_table}})\n\n<!--end-context-information-->\n<!--start-question-->\n\n**Question**: Hi How do I log audio using wandb?\n**Langauge**: en\n**Query Intents**: \n- The query is related to troubleshooting code using Weights & Biases\n- The query is related to a feature of Weights & Biases such as Sweeps, Artifacts, Reports, Experiments, Tables, Prompts, Weave, StreamTables and more\n\n<!--end-question-->\n<!--final-answer-in-markdown-->\n',
```
Comment (morganmcg1) [inline src/wandbot/rag/response_synthesis.py:99]: remove this old prompt?

File: src/wandbot/rag/utils.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    elif not isinstance(retrieval_input, str):
+        retrieval_input = str(retrieval_input)
+    return retrieval_input
+
+
+def get_web_contexts(web_results):
+    output_documents = []
+    if not web_results:
+        return []
+    # web_answer = web_results["web_answer"]
```
Comment (morganmcg1) [inline src/wandbot/rag/utils.py:99]: remove this code?

File: src/wandbot/rag/query_handler.py [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+    """An enhanced query"""
+
+    language: str = Field(
+        ...,
+        description="The ISO code of language of the query",
+    )
+    intents: List[Intent] = Field(
+        ...,
+        description=f"A list of one or more intents associated with the query. Here are the possible intents that "
+        f"can be associated with a query:\n{json.dumps(INTENT_DESCRIPTIONS, indent=2)}",
```
Comment (morganmcg1) [inline src/wandbot/rag/query_handler.py:147]: the extra whitespace should probably be removed in these descriptions if this is being turned into a prompt to GPT-4 right?

File: src/wandbot/rag/query_handler.py [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+    """A query for vector search"""
+
+    query: str = Field(
+        ...,
+        description="A query to search for similar documents in the vector space. This will be used to find documents "
+        "required to answer the query",
+    )
+
+
+class EnhancedQuery(BaseModel):
```
Comment (morganmcg1) [inline src/wandbot/rag/query_handler.py:137]: we're asking the model to do quite a lot here for a single call, in a future release we should consider breaking this up and making multiple calls in parallel that just do a single task each

**Review by morganmcg1 — APPROVED** [review]
lgtm, we should run evals on this before releasing it, given there are so maybe changes we should be fully aware of the impact of all these updates

**Comment by morganmcg1** [comment]
Oh also, can you add a Table in the README towards the top of with the history of the Answer Correctness score as we've moved through each release please

File: src/wandbot/ingestion/prepare_data.py (hunk 2 above)
Comment (parambharat) [reply src/wandbot/ingestion/prepare_data.py:765]: The reason for this was with the unhyphenated text it's easier to do metadata filter using `contains` which is not possible with the hyphenated text. The API can take tag filters than are unhyphenated.

File: src/wandbot/ingestion/prepare_data.py (hunk 3 above)
Comment (parambharat) [reply src/wandbot/ingestion/prepare_data.py:786]: same as above. It's easier to run contains query when using metadata filters with un-hyphenated strings.

File: src/wandbot/ingestion/prepare_data.py (hunk 4 above)
Comment (parambharat) [reply src/wandbot/ingestion/prepare_data.py:105]: removed

File: src/wandbot/rag/response_synthesis.py (hunk 10 above)
Comment (parambharat) [reply src/wandbot/rag/response_synthesis.py:99]: removed

File: src/wandbot/rag/utils.py (hunk 11 above)
Comment (parambharat) [reply src/wandbot/rag/utils.py:99]: removed

File: src/wandbot/rag/query_handler.py (hunk 12 above)
Comment (parambharat) [reply src/wandbot/rag/query_handler.py:147]: good catch, removed.

File: src/wandbot/rag/query_handler.py (hunk 13 above)
Comment (parambharat) [reply src/wandbot/rag/query_handler.py:137]: I tested out both parallel calls and this and found that a single call is not only faster, but also provides the model with context to fill in attributes with diversity.

File: src/wandbot/ingestion/vectorstores.py (hunk 8 above)
Comment (ayulockin) [reply src/wandbot/ingestion/vectorstores.py:62]: this is taken off in the wandbot-eval branch.

**Pull request closed without merging**